### ISSUE TITLE
OkHttpClient is final

### ISSUE THREAD

**Issue by williamokano** [OpenFeign/feign#729]
What is the point of `OkHttpClient` be a final class?

Some frameworks, like spring, make intense use of proxies for almost everything.

I had a use case where I use to build the feign clients manually and now I'm trying to use spring-cloud-opentracing to do some distributed tracing.

One problem I'm facing is, on that package there's a Aspect in order to decorate the client the their `TracingClient`, but the problem is: the aspect tries to make a proxy of the client implementation but the OkHttpClient is final.

tl;dr
What is the possibility of making the OkHttpClient **not** final?

**edit**:
The spring cloud opentracing feign can be found here: https://github.com/opentracing-contrib/java-spring-cloud/blob/master/instrument-starters/opentracing-spring-cloud-feign-starter/src/main/java/io/opentracing/contrib/spring/cloud/feign/FeignTracingAutoConfiguration.java#L80

**Comment by kdavisk6** [comment]
Based on my understanding of the history of this project and it's modules, the `Client` interface is meant to be "open for extension, closed for modification" following the **Open-Closed Principle**.  Both the `ApacheHttpClient` and the `OkHttpClient` are `final` in order to adhere to this.  

With regards to `OpenTracing`, there is an `OpenTracing` implementation for `Feign` available that appears to indicate that you can use the provided `OkHttpClient` with `OpenTracing`

https://github.com/OpenFeign/feign-opentracing

In your particular case, you want to use Spring's Open Tracing support for Spring Boot, which as you've pointed out, uses an `Aspect` to instrument a Feign client that is **not** using a `TracingClient` explicitly.  Since you have mentioned that you are manually configuring `Feign`, if you use a `TracingClient` explicitly, the `Aspect` is not necessary.

```java
GitHub github = Feign.builder()
                     .client(new TracingClient(new OkHttpClient(), tracer))
                     .target(GitHub.class, "https://api.github.com");
```

Looking at both `feign-opentracing` and the Spring project, if you want to use a provided `Client`, you must use `TracingClient` directly.

Now, with that said, I don't see an issue with changing the current `Client` implementations away from `final`.  Let's ask the other contributors what they think.

@velo, @spencergibb, @adriancole Thoughts?

**Comment by williamokano** [comment]
@kdavisk6 you're correct, the feign-opentracing provide a `TracingClient` to decorate another client.

On the other hand, the spring-boot-cloud-opentracing _should_ be able to autoconfigure all the needed beans, and on this case, the feign client, without having to explicitly set the tracing-client.

I respect the OCP from solid, and I think that standards are needed, but look at this class, it the implementation of the `RibbonClient` delegate for `Feign`, which is not final and works fine with the Aspect.

https://github.com/OpenFeign/feign/blob/master/ribbon/src/main/java/feign/ribbon/RibbonClient.java

As it is on the feign repository, I assume that it would not be a huge problem.

Well, if the answer is not changing, I might try to change the aspect pointcut to intercept the feign.setClient to "auto decorate" the client with the TracingClient.

**Comment by kdavisk6** [comment]
I spent some more time thinking about this, as I've been in your shoes on previous projects.  Specifically, trying to create an AutoConfiguration around a 3rd party library.  Assuming that the `Client` implementations remain `final`, you may be able to refactor `FeignTracingAutoConfiguration` to be **after** `FeignAutoConfiguration`, replacing the `Aspect` with an overridden  `FeignClient` with a `TracingClient`. 

```java
@Bean
public Client feignClient(FeignClient feignClient) {
   return new new TracingClientBuilder((Client) bean, tracer)
            .withFeignSpanDecorators(spanDecorators)
            .build()
}
```

I haven't tested this directly, but I have used similar techniques in the past to do this sort of thing.  This could work.

Alternatively, we can still remove the `final` modifier.

**Comment by williamokano** [comment]
I'm gonna try your solution.

Em qua, 4 de jul de 2018 17:16, Kevin Davis <notifications@github.com>
escreveu:

**Comment by williamokano** [comment]
Couldn't do as you suggested.

I created a springboot autoconfiguration which creates a bean for TracingClient decorated with OkHttpClient. I think it's probably for the best to keep OkHttpClient as final.
